### ISSUE TITLE
DTLS Handshake failure using Quectel BG77 module as a client

### ISSUE THREAD

**Issue by SB-Robin** [eclipse-californium/californium#2219]
Hello, I'm attempting to send a CoAP request over DTLS to our VPS server (with a static IPv4 address) using the Quectel BG77 module. However, each time I try to establish a handshake with the server, it fails with the error "DTLS handshaking failure" on client site. Interestingly, when testing with the Leshan sandbox server (leshan.eclipseprojects.io:5684), the communication works without any issues.

Environment:

Californium version: 3.11.0-SNAPSHOT
Cipher suite used: TLS_PSK_WITH_AES_128_CCM_8
Communication over CoAP (port 5683) with our server works fine
Attempting to connect to the cf-plugtest-server

I've also attempted to connect via OpenSSL. It was successful when connecting from the same VPS (localhost) as the cf-plugtest-server, but unsuccessful when attempting from a personal PC outside of the VPS network.
OpenSSL command used for both attempts: openssl s_client -dtls1_2 -cipher PSK-AES128-CCM8 -psk_identity Client_identity -connect 37.235.102.105:5684 -debug -psk 73656372657450534B

Both port 5683 and 5684 (for DTLS) are open on our VPS, and I've even tested with the firewall disabled, but the issue persists.

Could you please advise on potential reasons for the DTLS handshake failure with Californium in this setup? Thank you.

Successfull connect to Leshan server with PSK -> 864508030012428&313233343446473839
![successfull](https://github.com/eclipse-californium/californium/assets/132359529/08114ecd-995c-4e7f-a849-fda88674a0d4)

DTLS handshake failure when connecting cf-plugtest-server on our VPS with PSK -> Client_identity&secretPSK
![fail](https://github.com/eclipse-californium/californium/assets/132359529/9d667310-17c2-4008-ae98-8576a583e151)

**Comment by boaks** [comment]
> port 5683

Maybe better port 5684? If that doesn't help:

"our VPS Server"

[IP Capturing ‐ How To Provide The Right Information](https://github.com/eclipse-californium/californium/wiki/Logs-and-IP-Capturing-%E2%80%90-How-To-Provide-The-Right-Information#ip-capturing)

And, please captures, no pictures.

**Comment by boaks** [comment]
I get for "37.235.102.105" "Destination unreachable".

So, please try to create a capture on the server side.
Using tcpdump local on the server may already show, if traffic is incoming or not.

```
sudo tcpdump -i any 'udp port 5684'
```

**Comment by SB-Robin** [comment]
At that time I didn't have cf-plugtest-server enabled yet. After I started it, I typed "sudo tcpdump -i any 'udp port 5684'" and tried to send a handshake request from the module. The server only received this: 14:31:01.857228 eth0 In IP gw1-n1.cluster.ioteasyconnect.de.20156 > soh07.5684: UDP, length 107
[tcpdump.zip](https://github.com/eclipse-californium/californium/files/14195249/tcpdump.zip)

**Comment by boaks** [comment]
Just to mention:

```
openssl s_client -dtls1_2 -cipher PSK-AES128-CCM8 -psk_identity Client_identity -connect 37.235.102.105:5684 -debug -psk 73656372657450534B
```

Californium uses some [interoperability-tests](https://github.com/eclipse-californium/californium/tree/main/californium-tests/californium-interoperability-tests/src/test/java/org/eclipse/californium/interoperability/test/openssl), so if that doesn't work, please run the plugtest-server locally and test it again. If that succeeds (I think so), but in your cloud-box it doesn't, I guess there are some failure in the network config there.

**Comment by boaks** [comment]
Also to mention:

Californium runs also on a [sandbox](https://github.com/eclipse-californium/californium?tab=readme-ov-file#interop-server). If your client works with that, then this points pretty much to your cloud network setup.

**Comment by SB-Robin** [comment]
So I tried a second time, this time without the specified 0xC0A8 cipher suite and here is the result: 
[tcpdump.zip](https://github.com/eclipse-californium/californium/files/14205849/tcpdump.zip)
I don't know why 0xC0A8 cipher suite is not listed in the incoming suites, even though Quectel lists it... and it worked on the Leshan sandbox server before. 

@boaks  The server is on, feel free to try to connect. Should I also try to intercept the server side communication when you try it?

I don't know if I'm understanding you correctly, but I ran [openssl_server.sh](https://github.com/eclipse-californium/californium/blob/main/californium-tests/californium-interoperability-tests/openssl_server.sh) in one console and [openssl_client.sh](https://github.com/eclipse-californium/californium/blob/main/californium-tests/californium-interoperability-tests/openssl_client.sh) in the other console (for the local server) and it worked fine. But I don't know how to test it from my personal pc.

> Californium runs also on a [sandbox](https://github.com/eclipse-californium/californium?tab=readme-ov-file#interop-server). If your client works with that, then this points pretty much to your cloud network setup.

I tried to connect through the module to this sandbox (californium.eclipseprojects.io:5684) and I successfully connected only with 0xC0A8 cipher suite.

**Comment by boaks** [comment]
> I tried to connect through the module to this sandbox (californium.eclipseprojects.io:5684) and I successfully connected only with 0xC0A8 cipher suite.

So, once the network configuration is fixed, I'm looking forward to get your success report ;-).

> The server is on,

I still get "Destination unreachable" for "37.235.102.105". 

> I don't know if I'm understanding you correctly

I recommend to test using the openssl client with the sandbox and then with your server on your "local PC" (and use wireshark there to capture the client traffic):

```
openssl s_client -dtls1_2 -cipher PSK-AES128-CCM8 -psk_identity Client_identity -connect californium.eclipseprojects.io:5684 -debug -psk 73656372657450534B
```
(works on my machine)

```
openssl s_client -dtls1_2 -cipher PSK-AES128-CCM8 -psk_identity Client_identity -connect 37.235.102.105:5684 -debug -psk 73656372657450534B
```
(doesn't work on my machine)

[dtls_test.pcapng.gz](https://github.com/eclipse-californium/californium/files/14206453/dtls_test.pcapng.gz)

You may also check a server-side capture and compare that with your client side.

FMPOV all points to "network configuration" on the server side.

**Comment by SB-Robin** [comment]
Thank you very much for your promptness, patience and valuable advices. I appreciate it very much! :-) Finally it was successful and everything seems to be working. 

I would like to write what it was and how I managed to fix it, but I don't know myself. I tried following some advice on the internet and I don't think I even set up much. I just tried sending the handshake request again afterwards and it worked. :-)
